src/middleware/auth.js

```diff
@@ -0,0 +1,99 @@
+'use strict';
+
+'use strict';
+
+const jwt = require('jwt-simple');
+const wrap = require('co-express');
+const createError = require('http-errors');
+const _ = require('lodash');
+
+const EXPIRED = 'Token expired';
+const VALIDATION_FAILED = 'Token validation failed';
+const NOT_CONNECTED = 'User not connected';
+const NEED_KEY = 'IDP secret for JWT is not set';
+const ALGORITHM = 'RS256';
+const KEY_END_STRING = '-----END PUBLIC KEY-----\n';
+
+/**
+ * This functions takes a string of concatenated PEM public keys and splits
+ * them in an array.
+ * @param {string} keysString - Concatenated PEM public keys
+ * @returns {Array} Separated PEM public keys
+ */
+function retrieveKeys(keysString) {
+  const keys = keysString.split(KEY_END_STRING);
+
+  return _.map(keys, (item, index) => (index + 1 === keys.length) ? item : item + '-----END PUBLIC KEY-----');
+}
+
+/**
+ * This functions checks several PEM public keys among the
+ * ones defined in the keys array. This functions executes jwt.decode on
+ * every available PEM public keys and then returns the payload if there is a matching key.
+ * Otherwise it throws and error. Among these errors, the 'Token expired' one
+ * is threw in priority because it has more details than the 'Token validation failed' error.
+ * @param {array} keys - PEM public keys
+ * @param {string} token - Issuer JWT token
+ * @returns {object} The token's payload
+ * @throws Custom error with an HTTP code and message
+ */
+function validateToken(keys, token) {
+  let payload;
+  let error;
+
+  _.map(keys, key => {
+    try {
+      payload = jwt.decode(token, key, false, [ALGORITHM]);
+    } catch (err) {
+      if (err.message === EXPIRED) {
+        error = createError(401, EXPIRED);
+      }
+      if (_.get(error, 'message') !== EXPIRED) {
+        error = createError(401, VALIDATION_FAILED);
```

Comment:
This generic error handling is NOT OK :=)

This leads to the situation where there is a bug in our code, and we tell the user that his login / password is invalide. When this has reached production, it's really hard to realize what's happening, and it can last for months until someone we trust reports the problems (because a lot of users will type the wrong password, and swear that they typed it correctly)

What we need to do:

- Explicitely check for each of the errors that we can expect, and return the appropriate HTTP code (400, 401, 403, etc...à
- Treat a a bug any situation that we didn't expect: logger.error + return 500

This way of working will put us in a path where our solution can be improved over time.

If for security reason, we want to obfuscate the actual situation, we can do it (although I think it's not a good thing to do here), but we need to do it separately for each case, and still have that enumeration, with a catch all that consists in error + 500

Reply:
I totally agree and that's what came to my mind when I've seen how we dealt with the error handling in Gaia's auth middleware (the one I used to write this new one).

In fact I wanted to write a fine and detailed error handling but I'm absolutely not sure about how the services which are expecting authentication errors deals with the response. Are they well written enough to accept any kind of error response or are they hard written to only deal with some specific errors ? It needs some investigation

Reply:
Just a comment about this: 

I'm actually working on the same middleware in Go, to allow it to check with multiple public keys too, and it seems that it handles errors the same way Samy does ie returning 401 whatever the problem is https://github.com/transcovo/go-chpr-middlewares/blob/master/jwtAuth.go#L108